opencog/moses/scoring/precision_bscore.cc

```diff
@@ -123,325 +130,341 @@ precision_bscore::precision_bscore(const CTable& ctable_,
+		bs.push_back(0.5);
+	} else // Add 0.0 to ensure the bscore has the same size
+		bs.push_back(0.0);
+
+	// Append the activation penalty only if we are not performing
+	// boosting. Why? Because the boosted scorer only wants those
+	// selectors that are exactly correct, even if they have a terrible
+	// activation.  It will patch these together in the ensemble.  If
+	// these perfect scorers were penalized, they'd be discarded in the
+	// metapop, and we don't want that.  The member "_return_weighted_score"
+	// is a synonym for "we're doing boosting now".
+	if (not _return_weighted_score) {
+		// For boolean tables, activation sum of true and false positives
+		// i.e. the sum of all positives.   For contin tables, the activation
+		// is likewise: the number of rows for which the combo tree returned
+		// true (positive).
+		score_t activation = active / _ctable_weight;
+		score_t activation_penalty = get_activation_penalty(activation);
+		bs.push_back(activation_penalty);
+
+		if (logger().is_fine_enabled()) {
+			score_t precision = 0.0;
+			if (0 < active) {
+				// See above for explanation for the extra 0.5
+				precision = sao / active + 0.5;
+			}
+			logger().fine("precision = %f  activation=%f  activation penalty=%e",
+			              precision, activation, activation_penalty);
+		}
+
+		// Add time dispersion penalty
+		if (_pressure > 0.0) {
+			score_t dispersion_penalty =
+					get_time_dispersion_penalty(ctable_res.ordered_by_time());
+			bs.push_back(dispersion_penalty);
+			logger().fine("dispersion_penalty = %f", dispersion_penalty);
+		}
+	}
+
+	return bs;
 }
 
 /// scorer, for use with individual combo trees
-behavioral_score precision_bscore::operator()(const combo_tree& tr) const
+behavioral_score precision_bscore::operator()(const combo_tree &tr) const
+{
+	interpreter_visitor iv(tr);
+	auto interpret_tr = boost::apply_visitor(iv);
+
+	std::function<bool(const multi_type_seq &)> selector;
+	selector = [&](const multi_type_seq &irow) -> bool {
+		return id::logical_true == interpret_tr(irow.get_variant());
+	};
+
+	behavioral_score bs(do_score(selector));
+	log_candidate_bscore(tr, bs);
+	return bs;
+}
+
+behavioral_score precision_bscore::operator()(const Handle &handle) const
 {
-    interpreter_visitor iv(tr);
-    auto interpret_tr = boost::apply_visitor(iv);
-
-    std::function<bool(const multi_type_seq&)> selector;
-    selector = [&](const multi_type_seq& irow)->bool
-    {
-        return id::logical_true == interpret_tr(irow.get_variant());
-    };
-
-    behavioral_score bs(do_score(selector));
-    log_candidate_bscore(tr, bs);
-    return bs;
+	atomese::Interpreter interpreter(moses::compressed_value_key);
+	const ValuePtr result = interpreter(handle);
+	auto link_result = LinkValueCast(result)->value();
+	int i = 0;
+
+	std::function<bool(const multi_type_seq &)> selector;
+	selector = [&](const multi_type_seq &irow) -> bool {
+		bool result = bool_value_to_bool(link_result.at(i));
```

Comment:
It is slightly inelegant to use `do_score` without using `irow`, but yeah, I now understand why you didn't use `do_score`... Maybe we'll want to generalize `do_score` latter on but for now it's OK to let as it is, I think.